File: ui/visual-debugger/src/main/java/org/visual/debugger/context/AttachSceneContext.java
package org.visual.debugger.context;

import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.control.SplitPane;
import javafx.stage.Stage;
import lombok.Getter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public enum AttachSceneContext {
  INSTANCE;

  private static final Logger log = LoggerFactory.getLogger(AttachSceneContext.class);

  @Getter private static final SimpleObjectProperty<Stage> stage = new SimpleObjectProperty<>();
  @Getter private static final SimpleIntegerProperty nodeCounts = new SimpleIntegerProperty(0);

  static {
    stage.addListener(
        (observable, oldValue, newValue) -> {
          log.debug("Node count updated: {}", countNodes(newValue.getScene().getRoot()));
          nodeCounts.set(countNodes(newValue.getScene().getRoot()));
        });
  }

  private static int countNodes(Node node) {
    int count = 1;

    if (node instanceof Parent) {
      for (Node child : ((Parent) node).getChildrenUnmodifiable()) {
        count += countNodes(child); // 递归计算子节点
      }
    }

    if (node instanceof SplitPane) {
      log.warn("SplitPane detected: {}", ((SplitPane) node).getItems());
      for (Node child : ((SplitPane) node).getItems()) {
        count += countNodes(child);
      }
    }

    return count;
  }
}

File: ui/visual-debugger/src/main/kotlin/org/visual/debugger/context/AttachSceneContext.kt
package org.visual.debugger.context

import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.SplitPane
import org.visual.shared.KSlf4j

@KSlf4j
object AttachSceneContext {

    val scene by lazy { SimpleObjectProperty<Scene>() }

    val nodeCounts by lazy {
        SimpleIntegerProperty(0)
    }

    init {
        scene.addListener { _, _, newValue ->
            run {
                nodeCounts.set(countNodes(newValue.root))
            }
        }
    }

    private fun countNodes(node: Node): Int {
        var count = 1

        if (node is Parent) {
            for (child in node.childrenUnmodifiable) {
                count += countNodes(child) // 递归计算子节点
            }
        }

        if (node is SplitPane) {
            System.err.println(node.items)
            for (child in node.items) {
                count += countNodes(child)
            }
        }

        return count
    }
}